File: LRU_cache/LRU.java
import java.util.Map;
import java.util.HashMap;

public class LRU {
        public static void main(String[] args){
            problem p = new problem(6);
            p.put(5, 8);
            p.put(3,4);
            System.out.println(p.get(5));
            System.out.println(p.map);
        }
    }
    class problem{
            Node head = new Node(0,0);
            Node tail = new Node(0,0);
            Map<Integer, Node> map = new HashMap<>();
            int cap = 0;

            problem(int capacity){
                head.next = tail;
                tail.prev = head;
                cap = capacity;
            }

            public int get(int key){
                if(map.containsKey(key)){
                    Node temp = map.get(key);
                    remove(temp);
                    insert(temp);
                    return temp.val;
                } return -1;
            }

            public void put(int key, int val){  
                if(map.containsKey(key)){
                    remove(map.get(key));
                }
                if(map.size() == cap){
                    remove(tail.prev);
                }
                insert(new Node(key, val));
            }

            public void insert(Node node){
                map.put(node.key, node);
                node.next = head.next;
                node.next.prev = node;
                head.next = node;
                node.prev = head;
            }

            public void remove(Node node){
                map.remove(node.key);
                node.next.prev = node.prev;
                node.prev.next = node.next;
            }
        }

        class Node{
            Node prev, next;
            int key, val;
            Node(int k, int v){
                key = k;
                val = v;
        }
}

File: LRU_cache/LRU.py
class Node:
    def __init__(self, k ,v):
        self.key = k
        self.val = v
        self.next = None
        self.prev = None

class LRUCache(object):
    def __init__(self, capacity):
        self.cap = capacity
        self.map = {}
        self.head = Node(0,0)
        self.tail = Node(0,0)
        self.head.next = self.tail
        self.tail.prev = self.head
    
    def get(self, key):
        if key in self.map:
            temp = self.map[key]
            self.remove(temp)
            self.insert(temp)
            return temp.val
        return -1
    
    def put(self, key, value):
        if key in self.map:
            self.remove(self.map[key])
        if len(self.map) == self.cap:
            self.remove(self.tail.prev)
        self.insert(Node(key, value))


    def insert(self, node):
        self.map[node.key] = node
        node.next = self.head.next
        node.next.prev = node
        self.head.next = node
        node.prev = self.head
    
    def remove(self, node):
        del self.map[node.key]
        node.next.prev = node.prev
        node.prev.next = node.next

p = LRUCache(6)
p.put(3,4)
p.put(4,5)
print(p.get(4))